graphql/graphql.go

```diff
@@ -609,6 +604,7 @@ func (b *Block) resolveHeader(ctx context.Context) (*types.Header, error) {
 			b.header, err = b.r.backend.HeaderByHash(ctx, b.hash)
 		} else {
 			b.header, err = b.r.backend.HeaderByNumberOrHash(ctx, *b.numberOrHash)
+			b.hash = b.header.Hash()
```

Comment:
As far as I can tell, you moved setting `b.hash=xx` from oen location to another. How is the new location less racy?

Reply:
Ah

> This one is not really fixed imo, fixing it would introduce a mutex on t.block.numberOrHash which I can do, wanted to know some more opinions on it though.

> I just pushed a practical fix for the blockhash data race in graphql. We make sure to fill in block hash on initialization so they don't need to be fetched in the Logs resolver. However it goes to the slow-path if there was an edge case where hash wasn't filled in.

Reply:
Now it is less racy than when I wrote the comment you quote.

I moved setting the hash up the resolving tree. Now it is set by the parent thread before it spawns the child threads (Transaction objects) which race to set the hash. Because of this I think this code is less racy.